File: Top200/328-OddEvenList.java
package Top200;

class oddEvenList {
    class ListNode {
        int val;
        ListNode next;
        
        ListNode() {}
        
        ListNode(int val) {
            this.val = val;
        }

        ListNode(int val, ListNode next) {
            this.val = val;
            this.next = next;
        }
    }

    public ListNode oddEvenList(ListNode head) {
        ListNode dummy = new ListNode(-1);
        ListNode listOdd = new ListNode(-1);
        ListNode listEven = new ListNode(-1);

        ListNode curr = head;
        ListNode currOne = listOdd;
        ListNode currTwo = listEven;
        
        boolean odd = true; 
        
        while (curr != null) {
            if (odd) {
                currOne.next = curr;
                curr = curr.next;
                currOne = currOne.next;
                currOne.next = null;
            } else {
                currTwo.next = curr;
                curr = curr.next;
                currTwo = currTwo.next;
                currTwo.next = null;
            }
            
            odd = !odd;
        }

        dummy.next = listOdd.next;
        currOne.next = listEven.next;

        return dummy.next;
    }
    
}

File: Top200/328-OddEvenList.go
package main

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
type ListNode struct {
	Val  int
	Next *ListNode
}

func oddEvenList(head *ListNode) *ListNode {
	l1 := &ListNode{Val: -1, Next: nil}
	l2 := &ListNode{Val: -1, Next: nil}
	l1Bak := l1
	l2Bak := l2
	curr := head
	odd := true

	for curr != nil {
		if odd {
			l1.Next = curr
			l1 = l1.Next
			curr = curr.Next
			l1.Next = nil
		} else {
			l2.Next = curr
			l2 = l2.Next
			curr = curr.Next
			l2.Next = nil
		}

		odd = !odd
	}

	l1.Next = l2Bak.Next

	return l1Bak.Next
}
